src/IR/HM.hs

```diff
@@ -0,0 +1,490 @@
+{-# LANGUAGE DerivingVia #-}
```

Comment:
Can you include module documentation outlining the overall architecture of what's going on here/how to navigate this module?